modules/dhcp_common/free_ips.rb

```diff
@@ -0,0 +1,133 @@
+require 'concurrent'
+require 'logger'
+
+module Proxy::DHCP
+  class FreeIps
+    include ::Proxy::Log
+
+    DEFAULT_ADDRESS_KEEP_TIME = 30*60 # 30 mins
+    DEFAULT_CLEANUP_INTERVAL = 60 # 1 min
+
+    attr_reader :allocated_ips, :allocation_timestamps, :m
+
+    def initialize
+      @allocated_ips = []
+      @allocation_timestamps = []
+      @m = Monitor.new
+    end
+
+    # must be called from under the monitor
+    def mark_ip_as_allocated(ip_address)
+      @allocated_ips.push(ip_address)
+      @allocation_timestamps.push(time_now + DEFAULT_ADDRESS_KEEP_TIME)
+    end
+
+    def clean_up_allocated_ips
+      @m.synchronize do
+        loop do
+          logger.debug("Starting allocated ip addresses cleanup pass...")
+          break if @allocation_timestamps.first.nil? || @allocation_timestamps.first > time_now
+          @allocation_timestamps.shift
+          ip = @allocated_ips.shift
+          logger.debug("#{ip} marked as free.")
+        end
+      end
+    end
+
+    def start
+      logger.info("Starting allocated ip address maintenance (used by unused_ip call).")
+      @timer_task = Concurrent::TimerTask.new(:execution_interval => DEFAULT_CLEANUP_INTERVAL) { clean_up_allocated_ips }
+      @timer_task.execute
+    end
+
+    def stop
+      @timer_task.shutdown unless @timer_task.nil?
+    end
+
+    def time_now
+      Time.now.to_i
+    end
+
+    def find_free_ip(subnet_address, possible_ips, records)
+      @m.synchronize do
+        free_ips = possible_ips - @allocated_ips - records.collect{|record| record.ip}
```

Comment:
This line creates huge array in memory, it is very inefficient way picking an IP from a range. This test is a A class IP address range and the rake task is killed by OOM unable to finish the test case:

```
  def test_find_free_ip_aclass
    @subnet = Proxy::DHCP::Subnet.new("10.0.0.0", "255.0.0.0")
    @free_ips.stubs(:icmp_pingable?).returns(false)
    @free_ips.stubs(:tcp_pingable?).returns(false)
    r = Proxy::DHCP::Reservation.new('test', "10.0.0.1", "aa:bb:cc:dd:ee:ff", @subnet, :hostname =>'test')

    assert_equal "10.0.0.2", @free_ips.find_free_ip(@subnet.network, @subnet.valid_range(:from => "10.0.0.1", :to => "10.250.250.250"), [r])
  end
```

Although such a ranges are rather rare, this way of finding IPs is way too memory inefficient. When added with Ruby poor memory management and inability to return allocated memory to OS, this is an issue we should no longer accept in the codebase.